modules/api_test/tests/src/Functional/ApiTest.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Drupal\Tests\api_test\Functional;
+
+use Drupal\Component\Serialization\Json;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * @group headless
+ * @group api_test
+ */
+class ApiTest extends BrowserTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $profile = 'lightning_headless';
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = ['api_test'];
+
+  /**
+   * The access token returned by the api.
+   *
+   * @var string
+   */
+  protected $accessToken;
+
+  /**
+   * URL strings for different endpoints.
+   *
+   * @var string[]
+   */
+  protected $paths = [
+    'page_published_get' => '/jsonapi/node/page/api_test-published-page-content',
+    'page_unpublished_get' => '/jsonapi/node/page/api_test-unpublished-page-content',
+    'page_post' => '/jsonapi/node/page',
+    'role_get' => '/jsonapi/user_role/user_role',
+    'token_get' => '/oauth/token',
+  ];
+
+  /**
+   * Tests Getting data as anon and authenticated user.
+   */
+  public function testAllowed() {
+    // Get data that is available anonymously.
+    $client = \Drupal::httpClient();
+    $url = $this->buildUrl($this->paths['page_published_get']);
+    $response = $client->get($url);
+    $this->assertEquals(200, $response->getStatusCode());
+    $body = Json::decode($response->getBody());
+    $this->assertEquals('Published Page', $body['data']['attributes']['title']);
+
+    // Get data that requires authentication.
+    $client = \Drupal::httpClient();
+    $token = $this->getToken();
+    $url = $this->buildUrl($this->paths['page_unpublished_get']);
+    $options = [
+      'headers' => [
+        'Authorization' => 'Bearer ' . $token,
+        'Content-Type' => 'application/vnd.api+json'
+      ],
+    ];
+    $response = $client->get($url, $options);
+    $body = Json::decode($response->getBody());
+    $this->assertEquals('Unpublished Page', $body['data']['attributes']['title']);
+
+    // Post new content that requires authentication.
+    $count = count(\Drupal::entityQuery('node')->execute());
```

Comment:
Nit: This could be \Drupal::entityQuery('node')->count()->execute(). That returns an integer :)